src/test/java/com/puppycrawl/tools/checkstyle/JavaAstVisitorTest.java

```diff
@@ -0,0 +1,182 @@
+     */
+    private static final List<String> VISIT_METHODS_NOT_OVERRIDDEN = Arrays.asList(
+            // no tokens in production rule, so no AST to build
+            "visitClassOrInterfaceOrPrimitiveType",
+            "visitNonWildcardTypeArgs",
+            "visitStat",
+            "visitAnnotationConstantRest",
+            "visitSwitchLabeledRule",
+            "visitLocalVariableDeclaration",
+            "visitTypes",
+            "visitSwitchStat",
+            "visitSwitchPrimary",
+            "visitClassDef",
+            "visitInterfaceMemberDeclaration",
+            "visitMemberDeclaration",
+            "visitLiteralPrimary",
+            "visitPatternDefinition",
+            "visitLocalType",
+            "visitLocalTypeDeclaration",
+            "visitRecordBodyDeclaration",
+            "visitResource",
+            "visitVariableInitializer",
+            "visitLambdaBody",
+
+            // AST built in parent rule
+            "visitCreatedNameExtended",
+            "visitSuperSuffixSimple",
+            "visitFieldAccessNoIdent",
+            "visitClassType",
+            "visitClassOrInterfaceTypeExtended"
+    );
+
+    @Test
+    public void testAllVisitMethodsAreOverridden() {
+        final Method[] baseVisitMethods = CheckstyleJavaParserBaseVisitor
+                .class.getDeclaredMethods();
+        final Method[] visitMethods = JavaAstVisitor.class.getDeclaredMethods();
+
+        final Set<String> filteredBaseVisitMethodNames = Arrays.stream(baseVisitMethods)
+                .filter(method -> !VISIT_METHODS_NOT_OVERRIDDEN.contains(method.getName()))
+                .filter(method -> method.getName().contains("visit"))
+                .filter(method -> method.getModifiers() == Modifier.PUBLIC)
+                .map(Method::getName)
+                .collect(Collectors.toSet());
+
+        final Set<String> filteredVisitMethodNames = Arrays.stream(visitMethods)
+                .filter(method -> method.getName().contains("visit"))
+                .filter(method -> method.getModifiers() == Modifier.PUBLIC)
+                .map(Method::getName)
+                .collect(Collectors.toSet());
+
+        // remove overridden 'visit' method from ParseTreeVisitor interface in JavaAstVisitor
+        filteredVisitMethodNames.removeIf(name -> name.matches("visit"));
+
+        final String message = "Visit methods in 'CheckstyleJavaParserBaseVisitor' generated from "
+                + "production rules and labeled alternatives in 'CheckstyleJavaParser.g4' should "
+                + "be overridden in 'JavaAstVisitor' or be added to 'VISIT_METHODS_NOT_OVERRIDDEN' "
+                + "with comment explaining why.";
+
+        assertWithMessage(message)
+                .that(filteredVisitMethodNames)
+                .containsExactlyElementsIn(filteredBaseVisitMethodNames);
+    }
+
+    @Test
+        public void testOrderOfVisitMethodsAndProductionRules() throws Exception {
```

Comment:
I have determined that keeping production rules in `CheckstyleJavaParser.g4` and visit methods in `JavaAstVisitor.java` helps to ease maintenance, and should be considered a best practice. Example:

![image](https://user-images.githubusercontent.com/31252532/128538270-116f4312-12e6-4a85-91f4-35414e9e5000.png)


Just as we keep related production rules together in the grammar, so should we keep related `visit` methods together in `JavaAstVisitor.java`. This test makes sure that we keep these two files consistent.